### ISSUE TITLE
sslv3 Alert Handshake Failure (alert number 40)

### ISSUE THREAD

**Issue by gogo9th** [openssl/openssl#7147]
Hi,

I used "openssl s_client -connect <URL:port>" command to visit 500 most popular websites ([](https://moz.com/top500)). But some URLs cause the following error:

140271130755520:error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:../ssl/record/rec_layer_s3.c:1399:SSL alert number 40

The problematic hostnames are these (including the openssl command to connect to those HTTPS web servers):

openssl s_client -connect www.reuters.com:443 
openssl s_client -connect www.bloomberg.org:443
openssl s_client -connect www.wsj.com:443
openssl s_client -connect stumbleupon.com:443
openssl s_client -connect www-media.stanford.edu:443
openssl s_client -connect www.deviantart.com:443
openssl s_client -connect www.wiley.com:443
openssl s_client -connect themeforest.net:443
openssl s_client -connect mashable.com:443
openssl s_client -connect www.trustpilot.com:443
openssl s_client -connect www.a8.net:443
openssl s_client -connect www.uol.com.br:443
openssl s_client -connect www.domraider.com:443
openssl s_client -connect us.sagepub.com:443
openssl s_client -connect hbr.org:443
openssl s_client -connect static-service.prweb.com:443
openssl s_client -connect www.usgs.gov:443
openssl s_client -connect www.archives.gov:443
openssl s_client -connect www.usc.edu:443
openssl s_client -connect www.usa.gov:443
openssl s_client -connect www.istockphoto.com:443
openssl s_client -connect www.snapchat.com:443
openssl s_client -connect www2.gotomeeting.com:443
openssl s_client -connect bitbucket-marketing-cdn.atlassian.com:443
openssl s_client -connect cdn-1.wp.nginx.com:443
openssl s_client -connect d1af033869koo7.cloudfront.net:443
openssl s_client -connect www.worldbank.org:443
openssl s_client -connect www.mlbstatic.com:443

But when I use curl, none of the above causes any error:

curl -i https://www.reuters.com 
curl -i https://www.bloomberg.org
curl -i https://www.wsj.com handshake
curl -i https://stumbleupon.com
curl -i https://www-media.stanford.edu
curl -i https://www.deviantart.com
curl -i https://www.wiley.com
curl -i https://themeforest.net
curl -i https://mashable.com
curl -i https://www.trustpilot.com
curl -i https://www.a8.net
curl -i https://www.uol.com.br
curl -i https://www.domraider.com
curl -i https://us.sagepub.com
curl -i https://hbr.org
curl -i https://static-service.prweb.com
curl -i https://www.usgs.gov
curl -i https://www.archives.gov
curl -i https://www.usc.edu
curl -i https://www.usa.gov
curl -i https://www.istockphoto.com
curl -i https://www.snapchat.com
curl -i https://www2.gotomeeting.com
curl -i https://bitbucket-marketing-cdn.atlassian.com
curl -i https://cdn-1.wp.nginx.com
curl -i https://d1af033869koo7.cloudfront.net
curl -i https://www.worldbank.org
curl -i https://www.mlbstatic.com

My goal is to connect to the above web servers by using a nonblocking TLS. If I use a blocking TLS, I can connect to all of them fine. But if I use a non-blockching TLS, I get exactly the same error as connecting to the above web servers by using "opensssl."
 
I thought the issue was not setting the Hostname extension value in the SSL structure of my source code. So I put the correct hostname as the following:

SSL_set_tlsext_host_name(ssl, hostname);

However, I still get the same error.. If someone has a brilliant insight to solve this problem, please help me out. Thanks!

**Comment by kroeckx** [comment]
I have tried 5 random of those, they all work for me

**Comment by kroeckx** [comment]
Which version are you using?

**Comment by gogo9th** [comment]
My version is: OpenSSL 1.1.0g  2 Nov 2017.

If I run curl, it works. But openssl doesn't.

**Comment by gogo9th** [comment]
Okay, I figured out the cause of the problem. For certain web servers which have more than 1 hostname, the client has to tell the server the exact hostname the client is trying to connect to, so that the web server can present the right SSL certificate having the hostname the client is expecting. In case of blocking SSL, we should use "SSL_set_tlsext_host_name(SSL* ssl, char* hostname)" function to set the server's hostname. But if I use this function for my nonblocking SSL case, it doesn't work. The following is my code setting my client's SSL with the server's hostname:

char* hostname = "www.reuters.com";
SSL_CTX* ctx = SSL_CTX_new(TLS_method());
SSL* ssl = SSL_new(ctx);
BIO rbio = BIO_new(BIO_s_mem());
BIO wbio = BIO_new(BIO_s_mem());
SSL_set_bio(ssl, rbio, wbio); 
SSL_set_connect_state(ssl); 
**SSL_set_tlsext_host_name(ssl, hostname);**

Is it that the nonblocking SSL should use a different funcion than SSL_set_tlsext_host_name to set the server's hostname?

**Comment by kroeckx** [comment]
Blocking and non-blocking mode really work just the same, other
than that the sockets might block.

**Comment by gogo9th** [comment]
Okay, I figured out the full solution. When we use "openssl", if the connection gets terminated with "alert 40" error, that means we should explicitly specify the servername, so that the server can return the right certificate the client is expecting. For example, the following will work:

openssl s_client -connect www.domraider.com:443 **-servername www.domraider.com**

In my nonblocking OpenSSL, I simply made a mistake in my code. In other words, the "SSL_set_tlsext_host_name(ssl, hostname)" function is the right one the client should use to specify the server name.
